File: clients/android/app/src/main/kotlin/com/lextures/android/core/lms/ContentToolsObservability.kt
package com.lextures.android.core.lms

/** CT.M9 FR-17/18 — content-free mobile content-tools counters. */
object ContentToolsObservability {
    private val counters = mutableMapOf<String, Int>()
    private val lock = Any()
    const val PLATFORM = "android"

    fun record(event: String, toolId: String? = null, attributes: Map<String, String> = emptyMap()) {
        val attrs = attributes.toMutableMap()
        attrs["platform"] = PLATFORM
        if (!toolId.isNullOrBlank()) attrs["tool_id"] = toolId
        attrs["event"] = event
        if (!ContentToolGovernanceLogic.telemetryAttributesAreContentFree(attrs)) {
            // Drop payloads that would violate FR-18 rather than emit learner content.
            return
        }
        synchronized(lock) {
            val key = if (attrs.isEmpty()) {
                event
            } else {
                event + "|" + attrs.keys.sorted().joinToString(",") { "$it=${attrs[it]}" }
            }
            counters[key] = (counters[key] ?: 0) + 1
        }
    }

    fun count(event: String): Int = synchronized(lock) {
        counters.filter { it.key == event || it.key.startsWith("$event|") }.values.sum()
    }

    fun resetForTests() {
        synchronized(lock) { counters.clear() }
    }
}

File: clients/ios/Lextures/Core/LMS/ContentToolsObservability.swift
import Foundation

/// CT.M9 FR-17/18 — content-free mobile content-tools counters.
enum ContentToolsObservability {
    private static var counters: [String: Int] = [:]
    private static let lock = NSLock()
    static let platform = "ios"

    static func record(_ event: String, toolId: String? = nil, attributes: [String: String] = [:]) {
        var attrs = attributes
        attrs["platform"] = platform
        if let toolId, !toolId.isEmpty {
            attrs["tool_id"] = toolId
        }
        attrs["event"] = event
        guard ContentToolGovernanceLogic.telemetryAttributesAreContentFree(attrs) else {
            // Drop payloads that would violate FR-18 rather than emit learner content.
            return
        }
        lock.lock()
        defer { lock.unlock() }
        let key = attrs.isEmpty
            ? event
            : event + "|" + attrs.keys.sorted().map { "\($0)=\(attrs[$0] ?? "")" }.joined(separator: ",")
        counters[key, default: 0] += 1
    }

    static func count(for event: String) -> Int {
        lock.lock()
        defer { lock.unlock() }
        return counters.filter { $0.key == event || $0.key.hasPrefix(event + "|") }.values.reduce(0, +)
    }

    #if DEBUG
    static func resetForTests() {
        lock.lock()
        counters.removeAll()
        lock.unlock()
    }

    static func lastAttributes(for event: String) -> [String: String]? {
        lock.lock()
        defer { lock.unlock() }
        guard let key = counters.keys.first(where: { $0 == event || $0.hasPrefix(event + "|") }) else {
            return nil
        }
        guard let pipe = key.firstIndex(of: "|") else { return ["event": event] }
        let rest = key[key.index(after: pipe)...]
        var out: [String: String] = ["event": event]
        for part in rest.split(separator: ",") {
            let bits = part.split(separator: "=", maxSplits: 1).map(String.init)
            if bits.count == 2 { out[bits[0]] = bits[1] }
        }
        return out
    }
    #endif
}
